src/Guide/Api/Utils.hs

```diff
@@ -89,3 +94,43 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+
+-- | Type to log user's edition with header details
+data RequestDetails = RequestDetails
+  { rdIp      :: Text
+  , rdReferer :: Text
+  , rdUA      :: Text
+  } deriving (Show, Generic)
+
+instance (HasServer api context)
+  => HasServer (Maybe RequestDetails :> api) context where
+
+  type ServerT (Maybe RequestDetails :> api) m = Maybe RequestDetails -> ServerT api m
+
+  hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addHeaderCheck` withRequest headerCheck
+    where
+      headerCheck :: Request -> DelayedIO (Maybe RequestDetails)
+      headerCheck req = unfoldRequestArgument (Proxy :: Proxy '[Optional, Strict]) errReq errSt meRequestDatails
+        where
+          meRequestDatails :: Maybe (Either Text RequestDetails)
+          meRequestDatails = getCompose $ liftA2 RequestDetails (Compose ip) (Compose referer) <*> Compose ua
+
+          ip :: FromHttpApiData a => Maybe (Either Text a)
+          ip = fmap parseHeader $ lookup "X-Forwarded-For" (requestHeaders req)
+          referer :: FromHttpApiData a => Maybe (Either Text a)
+          referer = fmap parseHeader $ lookup hReferer (requestHeaders req)
+          ua :: FromHttpApiData a => Maybe (Either Text a)
+          ua = fmap parseHeader $ lookup hUserAgent (requestHeaders req)
```

Comment:
Maybe we case make something like:
```haskell
parseHeader :: String -> Maybe (Either Text a)
parseHeader key = lookup key (requestHeader req)

ip = parseHeader "X-Forwarded-For"
referer = parseHeader hReferer
ua = parseHeader hUserAgent
```

Also you can generalize RequestDetails over Text and then derive Traversable instance, so you just can write:
```haskell
meRequestDatails = getCompose $ traverse (Compose . parseHeader)
  (RequestDetails "X-Forwarded-For" hReferer hUserAgent)
```

Reply:
Simple example:

```haskell
{-# LANGUAGE DeriveFunctor, DeriveFoldable, DeriveTraversable #-}

import Data.Functor.Compose

data Error = Error
  deriving Show

data RequestDetails' a = RequestDetails
  { ip :: a
  , referer :: a
  , userAgent :: a
  } deriving (Functor, Foldable, Traversable, Show)

type RequestDetails = RequestDetails' String

parseHeader :: String -> Maybe (Either Error String)
-- parseHeader "ip" = Just (Left Error)
-- parseHeader "ip" = Nothing
parseHeader key = Just (Right key)

requestDetails :: Maybe (Either Error RequestDetails)
requestDetails = getCompose $ traverse (Compose . parseHeader) RequestDetails
  { ip = "ip"
  , referer = "referer"
  , userAgent = "userAgent"
  }
```

Reply:
I would advise against doing this because it relies on all fields having the same type. If we decide to parse IPs, for instance, it will get uglier.

Reply:
The reason it should be done right now is that the next thing I'm going to ask for in this PR is "please make the IP field have the type `IP`" 🙂

Reply:
By the way, the IP determining logic is more complicated than just taking the Forwarded-For header.

```haskell
  mbForwardedFor <- liftA2 (<|>) (Spock.header "Forwarded-For")
                                 (Spock.header "X-Forwarded-For")
  mbIP <- case mbForwardedFor of
    Nothing -> sockAddrToIP . Wai.remoteHost <$> Spock.request
    Just ff -> case readMaybe (toString ip) of
      Nothing -> error ("couldn't read Forwarded-For address: " ++
                        show ip ++ " (full header: " ++
                        show ff ++ ")")
      Just i  -> return (Just i)
      where
        addr = T.strip . snd . T.breakOnEnd "," $ ff
        ip -- [IPv6]:port
           | T.take 1 addr == "[" =
               T.drop 1 (T.takeWhile (/= ']') addr)
           -- IPv4 or IPv4:port
           | T.any (== '.') addr =
               T.takeWhile (/= ':') addr
           -- IPv6 without port
           | otherwise =
               addr
```

The "error" here is not nice, but everything else should be ported.

Reply:
Ok, it makes sense, it's a pity, I love Traversable

Reply:
Ok, it makes sense. It's a pity, I love Traversable.